cluster autoscaler is panicing with blank node taints added

I added some node taints like so:

```
k8s.io/cluster-autoscaler/node-template/taint/app |   |  
k8s.io/cluster-autoscaler/node-template/taint/group |  |
```

This is essentially a blank taint tag, for nodes/ASGs that aren't tainted.

I would expect the cluster autoscaler to handle this, but it panics and dies:

```
panic: runtime error: index out of range

goroutine 87 [running]:
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.extractTaintsFromAsg(0xc4219234a0, 0xc, 0xc, 0xc42196a180, 0xc42196a150, 0xc42021eaa0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_manager.go:314 +0x2e8
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.(*AwsManager).buildNodeFromTemplate(0xc4210b1200, 0xc420cda810, 0xc4215e7240, 0xc4215e7240, 0x0, 0x0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_manager.go:264 +0xb36
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.(*Asg).TemplateNodeInfo(0xc420cda810, 0xc4217a3c20, 0xc421125ad0, 0x22)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:355 +0x78
k8s.io/autoscaler/cluster-autoscaler/core.GetNodeInfosForGroups(0xc4215e7d80, 0x7, 0x8, 0x50bc120, 0xc420cda7b0, 0x50c64e0, 0xc420700ea0, 0xc4217a5e80, 0x4, 0x4, ...)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/utils.go:211 +0x65a
k8s.io/autoscaler/cluster-autoscaler/core.ScaleUp(0xc420e36b60, 0xc42179dbf0, 0x2, 0x2, 0xc4215e7d80, 0x7, 0x8, 0xc4217a5e80, 0x4, 0x4, ...)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/scale_up.go:57 +0x28e
k8s.io/autoscaler/cluster-autoscaler/core.(*StaticAutoscaler).RunOnce(0xc421547f10, 0xed2bc7ed7, 0xe056a939a, 0x518f0a0, 0x0, 0x0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/static_autoscaler.go:253 +0x13ea
k8s.io/autoscaler/cluster-autoscaler/core.(*PollingAutoscaler).RunOnce(0xc420d03300, 0xed2bc7ed7, 0xe056a939a, 0x518f0a0, 0x518f0a0, 0x4e200)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/polling_autoscaler.go:72 +0x146
main.run(0xc420a6c5a0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:262 +0x46f
main.main.func2(0xc4203509c0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:344 +0x2a
created by k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:145 +0x97
```

--- a/cluster-autoscaler/cloudprovider/aws/aws_manager.go
+++ b/cluster-autoscaler/cloudprovider/aws/aws_manager.go
@@ -22,6 +22,7 @@ import (
 	"fmt"
 	"io"
 	"math/rand"
+	"regexp"
 	"strings"
 	"time"
 
@@ -295,14 +296,18 @@ func extractTaintsFromAsg(tags []*autoscaling.TagDescription) []apiv1.Taint {
 	for _, tag := range tags {
 		k := *tag.Key
 		v := *tag.Value
-		splits := strings.Split(k, "k8s.io/cluster-autoscaler/node-template/taint/")
-		if len(splits) > 1 {
-			values := strings.SplitN(v, ":", 2)
-			taints = append(taints, apiv1.Taint{
-				Key:    splits[1],
-				Value:  values[0],
-				Effect: apiv1.TaintEffect(values[1]),
-			})
+		// The tag value must be in the format <tag>:NoSchedule
+		r, _ := regexp.Compile("(.*):NoSchedule")
+		if r.MatchString(v) {
+			splits := strings.Split(k, "k8s.io/cluster-autoscaler/node-template/taint/")
+			if len(splits) > 1 {
+				values := strings.SplitN(v, ":", 2)
+				taints = append(taints, apiv1.Taint{
+					Key:    splits[1],
+					Value:  values[0],
+					Effect: apiv1.TaintEffect(values[1]),
+				})
+			}
 		}
 	}
 	return taints

--- a/cluster-autoscaler/cloudprovider/aws/aws_manager.go
+++ b/cluster-autoscaler/cloudprovider/aws/aws_manager.go
@@ -297,16 +297,18 @@ func extractTaintsFromAsg(tags []*autoscaling.TagDescription) []apiv1.Taint {
 		k := *tag.Key
 		v := *tag.Value
 		// The tag value must be in the format <tag>:NoSchedule
-		r, _ := regexp.Compile("(.*):NoSchedule")
+		r, _ := regexp.Compile("(.*):(?:NoSchedule|NoExecute|PreferNoSchedule)")
 		if r.MatchString(v) {
 			splits := strings.Split(k, "k8s.io/cluster-autoscaler/node-template/taint/")
 			if len(splits) > 1 {
 				values := strings.SplitN(v, ":", 2)
-				taints = append(taints, apiv1.Taint{
-					Key:    splits[1],
-					Value:  values[0],
-					Effect: apiv1.TaintEffect(values[1]),
-				})
+				if len(values) > 1 {
+					taints = append(taints, apiv1.Taint{
+						Key:    splits[1],
+						Value:  values[0],
+						Effect: apiv1.TaintEffect(values[1]),
+					})
+				}
 			}
 		}
 	}

--- a/cluster-autoscaler/cloudprovider/aws/aws_manager_test.go
+++ b/cluster-autoscaler/cloudprovider/aws/aws_manager_test.go
@@ -70,15 +70,23 @@ func TestExtractTaintsFromAsg(t *testing.T) {
 			Value: aws.String("foo:NoSchedule"),
 		},
 		{
+			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/group"),
+			Value: aws.String("bar:NoExecute"),
+		},
+		{
+			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/app"),
+			Value: aws.String("fizz:PreferNoSchedule"),
+		},
+		{
 			Key:   aws.String("bar"),
 			Value: aws.String("baz"),
 		},
 		{
-			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/dedicated"),
+			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/blank"),
 			Value: aws.String(""),
 		},
 		{
-			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/dedicated"),
+			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/nosplit"),
 			Value: aws.String("some_value"),
 		},
 	}
@@ -89,10 +97,20 @@ func TestExtractTaintsFromAsg(t *testing.T) {
 			Value:  "foo",
 			Effect: apiv1.TaintEffectNoSchedule,
 		},
+		{
+			Key:    "group",
+			Value:  "bar",
+			Effect: apiv1.TaintEffectNoExecute,
+		},
+		{
+			Key:    "app",
+			Value:  "fizz",
+			Effect: apiv1.TaintEffectPreferNoSchedule,
+		},
 	}
 
 	taints := extractTaintsFromAsg(tags)
-	assert.Equal(t, 1, len(taints))
+	assert.Equal(t, 3, len(taints))
 	assert.Equal(t, makeTaintSet(expectedTaints), makeTaintSet(taints))
 }
 

--- a/cluster-autoscaler/cloudprovider/aws/aws_manager_test.go
+++ b/cluster-autoscaler/cloudprovider/aws/aws_manager_test.go
@@ -73,6 +73,14 @@ func TestExtractTaintsFromAsg(t *testing.T) {
 			Key:   aws.String("bar"),
 			Value: aws.String("baz"),
 		},
+		{
+			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/dedicated"),
+			Value: aws.String(""),
+		},
+		{
+			Key:   aws.String("k8s.io/cluster-autoscaler/node-template/taint/dedicated"),
+			Value: aws.String("some_value"),
+		},
 	}
 
 	expectedTaints := []apiv1.Taint{

